File: mobile/lib/core/loyalty/loyalty.dart
/// Một ưu đãi khách đủ điểm để đổi.
class Reward {
  const Reward({
    required this.rewardId,
    required this.name,
    this.description,
    required this.pointsRequired,
  });

  final String rewardId;
  final String name;
  final String? description;
  final int pointsRequired;

  factory Reward.fromJson(Map<String, dynamic> json) => Reward(
        rewardId: json['rewardId'] as String,
        name: json['name'] as String,
        description: json['description'] as String?,
        pointsRequired: json['pointsRequired'] as int,
      );
}

/// Điểm thưởng của chính tài khoản đang đăng nhập.
///
/// Ánh xạ `LoyaltyDtos.MyLoyaltyResponse`. Backend cố ý KHÔNG trả tổng chi tiêu — màn hình không
/// dùng tới, và trường nào không cần thì không gửi.
class MyLoyalty {
  const MyLoyalty({
    required this.linked,
    this.phoneNumber,
    required this.points,
    required this.availableRewards,
  });

  /// Tài khoản đã nối số điện thoại chưa.
  ///
  /// `false` là trạng thái BÌNH THƯỜNG của mọi tài khoản mới, không phải lỗi. Màn hình hiện lời
  /// mời liên kết chứ không hiện thông báo hỏng.
  final bool linked;

  final String? phoneNumber;
  final int points;
  final List<Reward> availableRewards;

  factory MyLoyalty.fromJson(Map<String, dynamic> json) => MyLoyalty(
        linked: (json['linked'] as bool?) ?? false,
        phoneNumber: json['phoneNumber'] as String?,
        points: (json['points'] as int?) ?? 0,
        availableRewards:
            ((json['availableRewards'] as List<dynamic>?) ?? const [])
                .map((e) => Reward.fromJson(e as Map<String, dynamic>))
                .toList(growable: false),
      );
}

File: mobile-rn/src/core/loyalty/loyalty.ts
/** Một ưu đãi khách đủ điểm để đổi. */
export interface Reward {
  readonly rewardId: string;
  readonly name: string;
  readonly description: string | null;
  readonly pointsRequired: number;
}

/**
 * Điểm thưởng của chính tài khoản đang đăng nhập.
 *
 * Ánh xạ `LoyaltyDtos.MyLoyaltyResponse`. Backend cố ý KHÔNG trả tổng chi tiêu — màn hình không
 * dùng tới, và trường nào không cần thì không gửi.
 */
export interface MyLoyalty {
  /**
   * Tài khoản đã nối số điện thoại chưa.
   *
   * `false` là trạng thái BÌNH THƯỜNG của mọi tài khoản mới, không phải lỗi. Màn hình hiện lời
   * mời liên kết chứ không hiện thông báo hỏng.
   */
  readonly linked: boolean;
  readonly phoneNumber: string | null;
  readonly points: number;
  readonly availableRewards: readonly Reward[];
}

/** Kết quả một lần đổi điểm (#34). */
export interface KetQuaDoiDiem {
  readonly redemptionId: string;
  readonly rewardName: string;
  readonly pointsSpent: number;
  /**
   * Số dư SAU khi đổi, do backend trả kèm.
   *
   * Không bắt app gọi thêm một lượt: sau khi tiêu điểm, con số khách muốn thấy ngay là số dư còn
   * lại, và một lượt gọi thứ hai tạo ra khoảng thời gian màn hình còn hiện số dư CŨ.
   */
  readonly soDuMoi: MyLoyalty;
}

export function rewardTuJson(json: unknown): Reward {
  const o = json as Record<string, unknown>;
  return {
    rewardId: o.rewardId as string,
    name: o.name as string,
    description: typeof o.description === 'string' ? o.description : null,
    pointsRequired: typeof o.pointsRequired === 'number' ? o.pointsRequired : 0,
  };
}

export function myLoyaltyTuJson(json: unknown): MyLoyalty {
  const o = (json ?? {}) as Record<string, unknown>;
  return {
    linked: typeof o.linked === 'boolean' ? o.linked : false,
    phoneNumber: typeof o.phoneNumber === 'string' ? o.phoneNumber : null,
    points: typeof o.points === 'number' ? o.points : 0,
    availableRewards: Array.isArray(o.availableRewards) ? o.availableRewards.map(rewardTuJson) : [],
  };
}

export function ketQuaDoiDiemTuJson(json: unknown): KetQuaDoiDiem {
  const o = json as Record<string, unknown>;
  return {
    redemptionId: typeof o.redemptionId === 'string' ? o.redemptionId : '',
    rewardName: typeof o.rewardName === 'string' ? o.rewardName : '',
    pointsSpent: typeof o.pointsSpent === 'number' ? o.pointsSpent : 0,
    soDuMoi: myLoyaltyTuJson(o.soDuMoi),
  };
}

/**
 * Khách đổi được ưu đãi này chưa.
 *
 * Tách thành hàm thuần vì đây là chỗ màn hình quyết định bật hay khoá nút, và cả hai điều kiện
 * đều bắt buộc: chưa liên kết số thì backend trả `LOYALTY_NOT_LINKED`, chưa đủ điểm thì trả
 * `LOYALTY_NOT_ENOUGH_POINTS`. Bật nút rồi để backend từ chối là bắt khách chạm vào một lời từ
 * chối lẽ ra thấy trước được.
 */
export function doiDuoc(diem: MyLoyalty, uuDai: Reward): boolean {
  return diem.linked && diem.points >= uuDai.pointsRequired;
}
